include/matx/transforms/solver.h

```diff
@@ -1529,21 +1525,21 @@ void eig_impl(OutputTensor &&out, WTensor &&w,
       detail::matxDnEigSolverPlan_t<OutputTensor, decltype(w_new), decltype(a_new)>::GetEigParams(w_new, tv, jobz, uplo);
 
   // Get cache or new eigen plan if it doesn't exist
-  auto ret = detail::dneig_cache.Lookup(params);
-  if (ret == std::nullopt) {
-    auto tmp = new detail::matxDnEigSolverPlan_t<OutputTensor, decltype(w_new), decltype(a_new)>{w_new, tv, jobz, uplo};
-
-    detail::dneig_cache.Insert(params, static_cast<void *>(tmp));
-    tmp->Exec(tv, w_new, tv, jobz, uplo, stream);
-  }
-  else {
-    auto eig_type =
-        static_cast<detail::matxDnEigSolverPlan_t<OutputTensor, decltype(w_new), decltype(a_new)> *>(ret.value());
-    eig_type->Exec(tv, w_new, tv, jobz, uplo, stream);
-  }
+  using cache_val_type = detail::matxDnEigSolverPlan_t<OutputTensor, decltype(w_new), decltype(a_new)>;
+  detail::GetCache().LookupAndExec<detail::eig_cache_t>(
+    detail::CacheName::EIG,
+    params,
+    [&]() {
+      return std::make_shared<cache_val_type>(w_new, tv, jobz, uplo);
+    },
+    [&](std::shared_ptr<cache_val_type> ctype) {
+      ctype->Exec(tv, w_new, tv, jobz, uplo, stream);
+    }
+  );
 
   /* Copy and free async buffer for transpose */
   matx::copy(out, tv.PermuteMatrix(), stream);
+  matxFree(tp);
```

Comment:
tp is used in tv, which is part of the construction of the plan eig_cache_t. do we only use tv for sizing , or is the op itself retained in any way? I just want to make sure we're not freeing memory here that the plan retains in some way.

Reply:
It's async-allocated memory, so it will only be freed after the copy to `out`